engine/src/main/java/nl/inl/blacklab/search/results/SpansReader.java

```diff
@@ -0,0 +1,334 @@
+
+    // Global counters, shared between instances of SpansReader in order to coordinate progress
+    final AtomicLong globalDocsProcessed;
+    final AtomicLong globalDocsCounted;
+    final AtomicLong globalHitsProcessed;
+    final AtomicLong globalHitsCounted;
+    /** Target number of hits to store in the {@link #globalResults} list */
+    final AtomicLong globalHitsToProcess;
+    /** Target number of hits to count, must always be >= {@link #globalHitsToProcess} */
+    final AtomicLong globalHitsToCount;
+    /** Master list of hits, shared between SpansReaders, should always be locked before writing! */
+    private final HitsInternalMutable globalResults;
+    /** Master list of capturedGroups (only set if any groups to capture. Should always be locked before writing! */
+    private CapturedGroupsImpl globalCapturedGroups;
+
+    // Internal state
+    boolean isDone = false;
+    private final ThreadAborter threadAborter = ThreadAborter.create();
+    private boolean isInitialized;
+    /* only valid after initialize() */
+    private int docBase; 
+
+    private boolean hasPrefetchedHit = false;
+    private int prevDoc = -1;
+
+    /**
+     * Construct an uninitialized SpansReader that will retrieve its own Spans object on when it's ran.
+     *
+     * HitsFromQueryParallel will immediately initialize one SpansReader (meaning its Spans object, HitQueryContext and
+     * CapturedGroups objects are set) and leave the other ones to self-initialize when needed.
+     *
+     * It is done this way because of an initialization order issue with capture groups.
+     * The issue is as follows:
+     * - we want to lazy-initialize Spans objects:
+     * 1. because they hold a lot of memory for large indexes.
+     * 2. because only a few SpansReaders are active at a time.
+     * 3. because they take a long time to setup.
+     * 4. because we might not even need them all if a hits limit has been set.
+     *
+     * So if we pre-create them all, we're doing a lot of upfront work we possibly don't need to.
+     * We'd also hold a lot of ram hostage (>10GB in some cases!) because all Spans objects exist
+     * simultaneously even though we're not using them simultaneously.
+     * However, in order to know whether a query (such as A:([pos="A.*"]) "ship") uses/produces capture groups (and how many groups)
+     * we need to call BLSpans::setHitQueryContext(...) and then check the number capture group names in the HitQueryContext afterwards.
+     *
+     * No why we need to know this:
+     * - To construct the CaptureGroupsImpl we need to know the number of capture groups.
+     * - To construct the SpansReaders we need to have already created the CaptureGroupsImpl, as it's shared between all of the SpansReaders.
+     *
+     * So to summarise: there is an order issue.
+     * - we want to lazy-init the Spans.
+     * - but we need the capture groups object.
+     * - for that we need at least one Spans object created.
+     *
```

Comment:
This is not an issue with this PR, but reading this, I'm wondering if it's possible to correctly construct the capture groups object from a single Spans instance.

A Spans corresponds to a segment, and a segment might not contain a term that does occur in the index as a whole. In this case, parts of the query may be optimized away, including potentially capture group clauses. This would mean the capture group name never gets registered if you're unlucky with your first Spans.

The solution would probably be to have it be possible to add group names later in the process if a new one is encountered.

Reply:
See https://github.com/INL/BlackLab/issues/353